[Chan Hon Boon Timothy] iP

timothysashimi opened the pull request:
## Toothless!

> _“What you’re searching for isn’t out there, hiccup. It’s in here.”_  🐉 

[Toothless](https://howtotrainyourdragon.fandom.com/wiki/Toothless_(Franchise)

Have you ever wanted a loyal companion like Toothless in HTTYD? Toothless has fortunately created a chatbot for you!

This chatbot is...
* User friendly
* ~Hard~ Super Easy to use
* Text Based

Here's a list of functions that Toothless can do for you:
1. Create Tasks for you
2. Store them
3. Fetch them according to keywords!

Here is a sample code block:
```java
public class Event inherits Task {
  public String taskName;

  Task(String taskName) {
    this.taskName = taskName;
  }
}
```


Here's what Toothless cannot do for you (yet):
- [x] Provide you with a GUI
- [ ] Remind you about tasks on certain dates

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

timothysashimi committed:
Level-0:

timothysashimi committed:
Merge pull request #1 from timothysashimi/week2

Level-0:

timothysashimi committed:
Level-1

timothysashimi committed:
Level-4

timothysashimi committed:
Merge pull request #2 from timothysashimi/week2

Week2 (level 4)

timothysashimi committed:
Level-5

timothysashimi committed:
Merge pull request #3 from timothysashimi/week2

Level-5

timothysashimi committed:
Level-6

timothysashimi committed:
Merge pull request #4 from timothysashimi/week2

Level-6

timothysashimi committed:
A-TextUiTesting

timothysashimi committed:
Merge pull request #5 from timothysashimi/week2

A-TextUiTesting

timothysashimi committed:
Level-7

timothysashimi committed:
Merge pull request #6 from timothysashimi/branch-Level-7

Level-7

timothysashimi committed:
Level-8

timothysashimi committed:
Merge pull request #7 from timothysashimi/branch-Level-8

Level-8

timothysashimi committed:
A-MoreOOP

timothysashimi committed:
A-Packages

timothysashimi committed:
Merge pull request #8 from timothysashimi/Week3A-

Week3A

timothysashimi committed:
Merge pull request #9 from timothysashimi/add-gradle-support

Add gradle support

timothysashimi committed:
A-Gradle

timothysashimi committed:
A-JUnit

timothysashimi committed:
A-JavaDoc

timothysashimi committed:
Merge pull request #10 from timothysashimi/Week3A-

A-JavaDoc

timothysashimi committed:
Level-9

timothysashimi committed:
Merge pull request #11 from timothysashimi/Level-9

Level-9

timothysashimi committed:
Level-10

timothysashimi committed:
Merge branch 'master' into Level-10

timothysashimi committed:
Merge pull request #12 from timothysashimi/Level-10

Level-10

timothysashimi committed:
Level-10

timothysashimi committed:
Level-10

timothysashimi committed:
Merge pull request #13 from timothysashimi/Level-10

Level-10

timothysashimi committed:
Added more javadocs

timothysashimi committed:
Better coding standards

chiaryan commented on src/main/java/duke/Exceptions/InvalidInstructionException.java:
```diff
@@ -0,0 +1,8 @@
+package duke.Exceptions;
```
I don't think this import statement follows the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#naming). The package name `Exceptions` should be in lowercase to differentiate from class names. I've noticed this in other parts of the PR as well.

chiaryan commented on src/main/java/duke/Parsers/DateTimeParser.java:
```diff
@@ -0,0 +1,52 @@
+     * @return DateTime object to be returned to store into a Task.
+     */
+    public static LocalDate stringToDT(String s) {
+        List<String> dateTimeFormats = Arrays.asList(
```
Good idea, having multiple formatters to capture the different ways the user inputs the date, might ~steal~ this for my own thing 😉

chiaryan commented on src/main/java/duke/Parsers/DateTimeParser.java:
```diff
@@ -0,0 +1,52 @@
+     * @param s String to be converted
+     * @return DateTime object to be returned to store into a Task.
+     */
+    public static LocalDate stringToDT(String s) {
```
Should this be `stringToDt` in order to follow camelcasing?

chiaryan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,147 @@
+import duke.Exceptions.MissingTaskToMarkException;
+import duke.Exceptions.MissingToDoNameException;
+import duke.Parsers.DateTimeParser;
+import duke.Tasks.*;
```
Should this import statement be expanded? So that you can keep the number of imported classes to a minimum

chiaryan commented on src/main/java/duke/Tasks/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public class Deadline extends Task {
+
+    String taskType;
```
The fields `taskType`, `taskName`, `isDone` are already defined in the parent class, not sure what is the purpose of redeclaring them here. Should these be removed? Same goes for the other `Task` subclasses.

chiaryan commented on src/main/java/duke/Tasks/Task.java:
```diff
@@ -0,0 +1,84 @@
+ */
+public class Task {
+    String taskName;
+    Boolean isDone;
```
Good job remembering to follow this part of the naming convention (`is...`). Seeing this made me realise I named my own booleans wrongly :P

chiaryan commented on src/main/java/duke/Tasks/TaskList.java:
```diff
@@ -0,0 +1,139 @@
+        output += t.toString() + "\n";
+        output += "Now you have " + tasksList.size() + " tasks in the list.";
+        return output;
+    }
```
I don't think it's necessary to have specialised method for each of the the Task subclasses. The `add(Task)` method would already capture all of the subclasses because of Java polymorphism. It would also make it easier to add other types of Tasks later.

chiaryan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,147 @@
+ */
+public class Parser {
+    //String response;
+    String line;
```
the `line` instance variable doesn't seem to change throughout the program runtime. Perhaps you could enforce the immutability by making it a `final` variable, or even make it a class-level member using `static`?

chiaryan reviewed (commented):
Awesum work :+1:. Couldn't find any major issues w.r.t. your coding standard, though I feel there can be a greater use of polymorphism in order to decrease code repetition.

timothysashimi committed:
Added assertions to the program

It is to document important assumptions that should hold at vaious points in the code. For example, the user entering text and pressing the enter button.

An AssertionError will be thrown and it will not be try and caught, indicating unrecoverable conditions. Let's handle edge cases in our program based on
the user's input text.

timothysashimi committed:
Merge branch 'master' into A-Assertions

timothysashimi committed:
Merge pull request #14 from timothysashimi/A-Assertions

Added assertions to the program

timothysashimi committed:
Refactored code to improve code quality

There are some violations previously, such as package names starting with capital letters.

In this commit, redundant methods were removed, package names were changed, and wildcard imports were removed.

timothysashimi committed:
Merge pull request #15 from timothysashimi/A-CodeQuality

Refactored code to improve code quality

timothysashimi committed:
Added streams in this commit

Previously, the find method uses a for loop to iterate through the tasklist. Hence a stream is introduced for efficiency.

timothysashimi committed:
Merge pull request #16 from timothysashimi/A-Streams

Added streams in this commit

timothysashimi committed:
Improved find method

Task name just has to match keyword partially for it to be recognised.

timothysashimi committed:
Merge pull request #17 from timothysashimi/A-CodeQuality

Improved find method

nuyer commented on src/main/java/duke/parsers/DateTimeParser.java:
```diff
@@ -0,0 +1,52 @@
+     * @param s String to be converted
+     * @return DateTime object to be returned to store into a Task.
+     */
+    public static LocalDate stringToDT(String s) {
```
can consider using "DateTime" here for this method instead or abbreviation to make your naming more consistent since you use DateTime (no acronym/abbreviation) in other places also!

nuyer commented on src/main/java/duke/parsers/FileParser.java:
```diff
@@ -0,0 +1,62 @@
+     * @param file File object to be passed
+     * @throws FileNotFoundException if file object is not found.
+     */
+    public void parseFile(File file) throws FileNotFoundException {
```
Very good that you specify what the method is parsing. Clear naming, no common mistakes 👍

nuyer commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,146 @@
+     * @return Returns a String response and modified TaskList
+     * @throws InvalidInstructionException if instruction is not formatted properly or has logical issues.
+     */
+    public Pair<TaskList, String> parse(TaskList tasksList, String input) throws InvalidInstructionException {
```
Czn be more specific on naming (specify what you are parsing)!

nuyer commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,54 @@
+     * @return returns a modified taskList for further modifications during the run function.
+     */
+
+    public TaskList printMessage(String message, Parser parser, TaskList taskList) {
```
Good and clear naming! using verb for method 🔥

nuyer commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Event extends Task {
+
+    String taskType;
```
good naming for fields/variable

nuyer commented on src/main/java/duke/tasks/TaskList.java:
Good naming for methods, all are verbs, following coding standard/ improve code quality GREAT JOB 🆙

nuyer reviewed (commented):
Your code is well written and structured. Most the variable/method names follow good coding practice. GREAT JOB 💯

timothysashimi committed:
Added more Java Docs

timothysashimi committed:
Merge branch 'master' of https://github.com/timothysashimi/ip into A-Assertions

timothysashimi committed:
Merge branch 'master' of https://github.com/timothysashimi/ip into A-JavaDoc

timothysashimi committed:
Added more javadoc comments

timothysashimi committed:
Merge pull request #19 from timothysashimi/heads/A-JavaDoc

Heads/a java doc

timothysashimi committed:
Merge branch 'master' into heads/A-CodingStandard

timothysashimi committed:
Merge pull request #21 from timothysashimi/heads/A-CodingStandard

Better coding standards

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Merge pull request #22 from timothysashimi/Published-UG

Update README.md

timothysashimi committed:
Some text formatting changes

timothysashimi committed:
Merge https://github.com/timothysashimi/ip

timothysashimi committed:
Added Ui.png

timothysashimi committed:
Merge pull request #23 from timothysashimi/Ui.png

Added Ui.png

timothysashimi committed:
Refactored code by shifting the test classes into packages.

Previously, the test classes were in the package test, which did not mirror the file structure of the duke class.
Hence, new packages were created and the test classes TaskListTest and ToDoTest were shifted into new packages under the test package.

timothysashimi committed:
Adhere to SLAP best practices.

Previously, the parse method was very long as executing each command was
not segregated into different methods.

Lets encapsulate each command execution into separate methods to delegate
responsibility to different methods.

timothysashimi committed:
Merge branch 'master' into A-Assertions

timothysashimi committed:
Merge pull request #18 from timothysashimi/A-Assertions

Add more Java Docs

Previously, some methods did not have proper documentation.

Let's add more documentation to methods so readers can have a better understanding of the functionalities.

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update READme.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Update README.md

timothysashimi committed:
Merge branch 'master' of https://github.com/timothysashimi/ip

timothysashimi committed:
Improve storage for Jar

timothysashimi committed:
Merge pull request #24 from timothysashimi/Fix-Storage

Improve storage for Jar
